test/functional/ui/decorations_spec.lua

```diff
@@ -1634,4 +1634,87 @@ l5
 
   end)
 
+  it('can add lots of signs', function()
+    screen:try_resize(40, 10)
+      helpers.command 'normal 10oa b c d e f g h'
+
+    for i = 1, 10 do
+      meths.buf_set_extmark(0, ns, i,  0, { end_col =  1, hl_group='Todo' })
+      meths.buf_set_extmark(0, ns, i,  2, { end_col =  3, hl_group='Todo' })
+      meths.buf_set_extmark(0, ns, i,  4, { end_col =  5, hl_group='Todo' })
+      meths.buf_set_extmark(0, ns, i,  6, { end_col =  7, hl_group='Todo' })
+      meths.buf_set_extmark(0, ns, i,  8, { end_col =  9, hl_group='Todo' })
+      meths.buf_set_extmark(0, ns, i, 10, { end_col = 11, hl_group='Todo' })
+      meths.buf_set_extmark(0, ns, i, 12, { end_col = 13, hl_group='Todo' })
+      meths.buf_set_extmark(0, ns, i, 14, { end_col = 15, hl_group='Todo' })
+      meths.buf_set_extmark(0, ns, i, -1, { sign_text='W' })
+      meths.buf_set_extmark(0, ns, i, -1, { sign_text='X' })
+      meths.buf_set_extmark(0, ns, i, -1, { sign_text='Y' })
+      meths.buf_set_extmark(0, ns, i, -1, { sign_text='Z' })
+    end
+
+    screen:expect{grid=[[
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:h}                 |
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:h}                 |
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:h}                 |
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:h}                 |
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:h}                 |
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:h}                 |
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:h}                 |
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:h}                 |
+      X Y Z W {3:a} {3:b} {3:c} {3:d} {3:e} {3:f} {3:g} {3:^h}                 |
+                                              |
+    ]], attr_ids={
+      [1] = {background = Screen.colors.Grey, foreground = Screen.colors.DarkBlue};
+      [2] = {bold = true, foreground = Screen.colors.Blue1};
+      [3] = {background = Screen.colors.Yellow, foreground = Screen.colors.Blue1};
+    }}
+  end)
+
+end)
+
+describe('decorations: virt_text', function()
+  it('avoids regression in #17638', function()
+    clear()
+
+    local screen screen = Screen.new(50, 10)
+    screen:attach()
+
+    exec_lua[[
+      vim.wo.number = true
+      vim.wo.relativenumber = true
+
+      vim.cmd 'normal 4ohello'
+      vim.cmd 'normal aVIRTUAL'
+    ]]
+
+    local ns = meths.create_namespace('test')
+
+    meths.buf_set_extmark(0, ns, 2, 0, {
+      virt_text = {{"hello", "String"}},
+      virt_text_win_col = 20,
+    })
+
+    -- Trigger a screen update
```

Comment:
I'd suggest adding a `screen:expect` before this. and then just do `'j'` here so we test the effect of both a normal screen update and then of a "number only" update.

Reply:
I did `kj`, so I could re-use the same grid.